src/values/AbstractValue.js

```diff
@@ -85,11 +81,65 @@ export default class AbstractValue extends Value {
   args: Array<Value>;
   _buildNode: void | AbstractValueBuildNodeFunction | BabelNodeExpression;
 
-  equals(x: any) {
+  addSourceLocationsTo(locations: Array<BabelNodeSourceLocation>) {
+    if (this._buildNode && !(this._buildNode instanceof Function)) {
+      if (this._buildNode.loc) locations.push(this._buildNode.loc);
+    }
+    for (let val of this.args) {
+      if (val instanceof AbstractValue) val.addSourceLocationsTo(locations);
+    }
+  }
+
+  addSourceNamesTo(names: Array<string>) {
+    let gen = this.$Realm.preludeGenerator;
+    function add_intrinsic(name: string) {
+      if (name.startsWith("_$")) {
+        if (gen === undefined) return;
+        add_args(gen.derivedIds.get(name));
+      } else if (names.indexOf(name) < 0) {
+        names.push(name);
+      }
+    }
+    function add_args(args: void | Array<Value>) {
+      if (args === undefined) return;
+      for (let val of args) {
+        if (val.intrinsicName) {
+          add_intrinsic(val.intrinsicName);
+        } else if (val instanceof AbstractValue) {
+          val.addSourceNamesTo(names);
+        } else if (val instanceof StringValue) {
+          if (val.value.startsWith("__")) {
+            names.push(val.value.slice(2));
+          }
+        }
+      }
+    }
+    if (this.intrinsicName) {
+      add_intrinsic(this.intrinsicName);
+    }
+    add_args(this.args);
+  }
+
+  buildNode(args: Array<BabelNodeExpression>): BabelNodeExpression {
+    let buildNode = this.getBuildNode();
+    return buildNode instanceof Function
+      ? ((buildNode: any): AbstractValueBuildNodeFunction)(args)
+      : ((buildNode: any): BabelNodeExpression);
+  }
+
+  equals(x: Value) {
+    if (x instanceof ConcreteValue) {
+      if (this.values._elements !== undefined && this.values._elements.length === 1) {
```

Comment:
Is this for template objects? Otherwise, why do we allow abstract values with a value domain of size 1?

Reply:
Right now, this should never arise. I'll remove it.